File: python/1599. Maximum Profit of Operating a Centennial Wheel.py
class Solution(object):
    def minOperationsMaxProfit(self, customers, boardingCost, runningCost):
        """
        :type customers: List[int]
        :type boardingCost: int
        :type runningCost: int
        :rtype: int
        """
        waiting_num = 0
        in_num = 0
        max_profit = 0
        rd = 0
        ans = -1
        while True:
            if rd < len(customers):
                waiting_num += customers[rd]
            in_num += min(4, waiting_num)
            waiting_num = max(0, waiting_num - 4)
            cur_profit = in_num*boardingCost - (rd+1)*runningCost
            if cur_profit > max_profit:
                max_profit = cur_profit
                ans = rd + 1
            rd += 1
            if rd >= len(customers) and waiting_num <= 0:
                break

        return ans

File: go/src/leetcode/1599/1599. Maximum Profit of Operating a Centennial Wheel.go
package main

/*
*
https://leetcode.cn/problems/maximum-profit-of-operating-a-centennial-wheel/description/
*/
func minOperationsMaxProfit(customers []int, boardingCost int, runningCost int) (ans int) {
	ans = -1
	var inNum, waitNum, maxProfit, curProfit int
	n := len(customers)
	for rd := 0; ; rd++ {
		if rd < n {
			waitNum += customers[rd]
		}
		inNum += min(4, waitNum)
		waitNum = max(0, waitNum-4)
		curProfit = inNum*boardingCost - runningCost*(rd+1)
		if curProfit > maxProfit {
			maxProfit = curProfit
			ans = rd + 1
		}
		if rd >= n && waitNum <= 0 {
			break
		}
	}
	return
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
